oauth_dropins/bluesky.py

```diff
@@ -0,0 +1,76 @@
+from atproto.exceptions import UnauthorizedError
+from . import views, models
+from flask import request
+from google.cloud import ndb
+from .webutil.util import json_dumps
+import atproto
+
+class Start(views.Start):
+  """
+  Starts Bluesky auth - used just to provide the button.
+  """
+  NAME = 'bluesky'
+  LABEL = 'Bluesky'
+
+class BlueskyAuth(models.BaseAuth):
+  """
+  An authenticated Bluesky user.
+  """
+
+  did = ndb.StringProperty(required=True)
+  password = ndb.StringProperty(required=True)
+  user_json = ndb.TextProperty(required=True)
+
+  def site_name(self):
+    return 'Bluesky'
+
+  def access_token(self):
+    """
+    Returns did and password as a tuple in place of an OAuth key id/secret pair.
+    """
+    return (self.did, self.password)
+
+  def _api(self):
+    """
+    Returns an atproto.Client.
+    """
+    (client, _) = BlueskyAuth._api_from_password(self.did, self.password)
+    return client
+
+  @staticmethod
+  def _api_from_password(username, password):
+    client = atproto.Client()
+    profile = client.login(username, password)
+    return (client, profile)
+
+
+class Callback(views.Callback):
+  """
+  OAuth callback stub.
+  """
+  def dispatch_request(self):
+    username = request.values.get('username')
+    password = request.values.get('password')
+
+    # get the did (portable user ID)
+    try:
+      (client, profile) = BlueskyAuth._api_from_password(username, password)
+    except UnauthorizedError:
+      return self.finish(None)
+    user_json = json_dumps({
+      '$type': profile._type,
+      'did': profile.did,
+      'handle': profile.handle,
+      'avatar': profile.avatar,
+      'banner': profile.banner,
+      'displayName': profile.displayName,
+      'description': profile.description,
+    })
+    auth = BlueskyAuth(
+      id=profile.handle,
```

Comment:
should we use the DID as the id instead?

Reply:
No reason not to from this side but isn't that what ends up in the URL? Which would then need to be URL-encoded and stuff as DIDs contain colons. If that's not the case then yeah DID would definitely be better! I'll have a play with it locally

Reply:
Oh no, it's fine actually, DIDs are OK in URLs. Just depends on whether you're ok with the aesthetics of it!

<img width="454" alt="image" src="https://github.com/snarfed/oauth-dropins/assets/2765356/cd290a59-6069-479d-9c48-375c265cd28f">

Reply:
Hah yeah not ideal. Technically a Bridgy thing, not oauth-dropins, but still. Bridgy does support a number of URL shortcuts, eg using domain like https://brid.gy/mastodon/joelotter.com redirects to https://brid.gy/mastodon/@joel@otter.garden , but yeah the id is still the canonical URL. Meh. 🤷